### ISSUE TITLE
Edge index in batch DataLoader

### ISSUE THREAD

**Issue by vymao** [rusty1s/pytorch_geometric#1827]
## ❓ Questions & Help

Does the Pytorch Geometric DataLoader keep track of which edge indices in the batch object belong to which graph? I ask because I am wondering whether or not the node indices must be distinct between graph Data objects.

**Comment by rusty1s** [comment]
Currently, the batching procedure expects nodes belonging to distinct graphs. You can, nonetheless, find out which edge belongs to which graph via
```python
row, col = edge_index
edge_batch = batch[row]
```

**Comment by vymao** [comment]
I'm a little confused. This seems like it depends on the size of the batch, and that will determine the size of `data.batch`. If this implies that different nodes between different graphs must have distinct indices, then wouldn't the indices for some nodes in some batch objects be more than the size of the batch? 

For example, let's say I had 4 Data objects with two nodes in each graph. I could label them as [(0,1), (0,1), (0,1), (0,1)] or [(1,2), (3,4), (5,6), (7,8)]. But I could batch graphs 1 and 3 together, creating a size of 4, but then the first notation would give me the wrong corresponding graph, and the second notation would have the (5,6) indices as too large.

**Comment by rusty1s** [comment]
In case you have 4 data objects with 2 nodes each, you should label them as `[(0, 1), (0, 1), (0, 1), (0, 1)]`. Indices are automatically increased when batching graphs together. In this case, batching all graphs together would let to a single data object with 8 nodes, labeled as `(0, 1, 2, 3, 4, 5, 6, 7)`.

I'm not sure what you mean with giving you "the wrong graph". In case you do not want to treat distinct graphs that way, you can overwrite the `collate_fn` in `torch.utils.data.DataLoader` and implement your own batching procedure, *e.g.*:
```python
def collate_fn(data_list):
    # Return a single `Batch` object out of a list of `Data` objects

loader = torch.utils.data.DataLoader(dataset, batch_size=..., collate_fn=collate_fn)
```

**Comment by akpas001** [comment]
yeah. of course. Earlier, I was doing that but this is how it looks when I remove unsqueeze. the whole data is wrong now. I am supposed to have 
data = Batch(x=[9, 30], edge_index=[2, 20], y=[9, 30], action=[1, 30], rewards=[1], active_edges=[1], batch=[32], ptr=[33]).

![1](https://user-images.githubusercontent.com/56883247/136666734-e8b6ba2a-7dbb-4222-a3df-7743c74923f5.PNG)
![2](https://user-images.githubusercontent.com/56883247/136666735-1b06522b-77d4-47eb-bf27-d210586a903a.PNG)
![3](https://user-images.githubusercontent.com/56883247/136666736-258c39c0-3b9b-465c-aa5b-5c36afed3d73.PNG)

but instead having the shape as shown in the image attached above. The batch size is getting multiplied with the num_nodes, num_edges. I do not understand why?


![4](https://user-images.githubusercontent.com/56883247/136666880-ed48b114-ece4-4443-9d24-26f6db10b1d6.PNG)
As per the above image, the data is appending perfectly fine to the dataset. But the moment it enters the loader it is getting messed up.

**Comment by akpas001** [comment]
Hello, for the sake of understanding I have created a dummy representation of how my data is being processed. 
[github_upload.txt](https://github.com/pyg-team/pytorch_geometric/files/7329502/github_upload.txt). So, I get my x and edge index from an experience_buffer, and from that, I am supposed to extract the x and edge_index and then feed them into my network. 

In this example : 

each x has 9 nodes and 30 features
each edge_index is of the shape (2,8)
 
since I have other attributes also to pass into the network(which I have not included here), I felt it is good to use a DataLoader. 
so, what happening here is, the batch_size is getting multiplied with the elements and producing some random data which doesn't make any sense. I would like to overcome the issue.

**Comment by shernankid** [comment]
> Currently, the batching procedure expects nodes belonging to distinct graphs. You can, nonetheless, find out which edge belongs to which graph via
> 
> ```python
> row, col = edge_index
> edge_batch = batch[row]
> ```

Hi Matthias,

I have the following questions.
1. where is the batch in "batch[row]" defined?  
2. I already processed the pytorch geometric dataset, is it possible to know which edges belong to which data in MessagePassing? (I created a class, class Encoder(MessagePassing))

Best,
shernan

**Comment by rusty1s** [comment]
Can you clarify? `batch[row]` will just map the node-level batch indices to edge-level ones. This is pure PyTorch code.

**Comment by purvanshi** [comment]
@rusty1s batch[row] will just map the node-level batch indices to edge-level ones. This is pure PyTorch code.
Is batch from the iterator here i.e for batch in dataloader:

**Comment by rusty1s** [comment]
Yes, that's correct. `batch` is the auto-generated vector from `DataLoader`.
